swri_route_util/src/util.cpp

```diff
@@ -373,4 +377,140 @@ bool interpolateRoutePosition(RoutePoint &dst,
                           norm_position.distance);
   return true;
 }
+
+bool routeDistance(
+  double &distance,
+  const mnm::RoutePosition &start,
+  const mnm::RoutePosition &end,
+  const Route &route)
+{
+  size_t start_index;
+  if (!route.findPointId(start_index, start.id)) {
+    return false;
+  }
+
+  size_t end_index;
+  if (!route.findPointId(end_index, end.id)) {
+    return false;
+  }
+
+  size_t min_index = std::min(start_index, end_index);
+  size_t max_index = std::max(start_index, end_index);
+
+  double d = 0.0;
+  if (route.header.frame_id == stu::_wgs84_frame) {
+    for (size_t i = min_index; i < max_index; i++) {
+      d += stu::GreatCircleDistance(route.points[i+1].position(), route.points[i].position());
+    }
+  } else {
+    for (size_t i = min_index; i < max_index; i++) {
+      d += (route.points[i+1].position() - route.points[i].position()).length();
+    }
+  }
+
+  if (end_index < start_index) {
+    d = -d;
+  }
+
+  distance = d + end.distance - start.distance;
+  return true;
+}
+
+bool routeDistances(
+  std::vector<double> &distances,
+  const mnm::RoutePosition &start,
+  const std::vector<mnm::RoutePosition> &ends,
+  const Route &route)
+{
+  size_t start_index;
+  if (!route.findPointId(start_index, start.id)) {
+    // Without a start index, we can't calculate anything.
+    return false;
+  }
+
+  // Find the indices for every point in the route, and the minimum
+  // and maximum indices that we need to work over (the ROI).
+  size_t min_index = start_index;
+  size_t max_index = start_index;
+  std::vector<int> indices;
+  indices.resize(ends.size());
+  for (size_t i = 0; i < ends.size(); ++i) {
+    size_t pt_index;
+    if (route.findPointId(pt_index, ends[i].id)) {
+      indices[i] = pt_index;
+      min_index = std::min(min_index, pt_index);
+      max_index = std::max(max_index, pt_index);
+    } else {
+      indices[i] = -1;
+    }
+  }
+
+  // This is the index of the start point in the ROI.
+  const size_t roi_start_index = start_index - min_index;
+
+  // We calculate the arc length of each point in the ROI relative to
+  // the start point.  This vector covers the ROI (so it corresponds
+  // from min_index to max_index)
+  std::vector<double> arc_lengths;
+  arc_lengths.resize(max_index-min_index+1);
+
+  // arc_lengths[0] = 0.0;
```

Comment:
Good catch.  That was the straightforward version that I did as a reference for writing the actual version I ended up with.  Added commits to both merge requests to remove it.